### PR TITLE
Fix scenario crash and random year (#45901)

### LINKED ISSUE #45901: Scenarios with explicit winter start actually start "whenever". (Including years before the charater was even born.)
# Describe the bug

Tried starting with "ambush" scenario and "sheltered" scenario. 
Both end up flinging you into random season of a random year. Or just crashing outright on world generation.
And I do mean **completely** random. Like, "years before the character was even born" random.
And it's not even winter.
 
Maybe or may not be related to https://github.com/CleverRaven/Cataclysm-DDA/pull/45466

# Steps To Reproduce

1) Create a fresh world in latest experimental. 
2) Start with "ambush" scenario and "sheltered" scenario. 
3) Look at the time.

# Expected behavior

You start during the winter of the first year. 

# Screenshots

random_year
Default bionic prepper start.

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 2004
- Game Version: 0.E-8087-g643d8c2 [32-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions],
    Bionic Slots [cbm_slots]
]

### PR DESCRIPTION
**Pull request by satheon49**
#### Summary

SUMMARY: Bugfixes "Fixes scenarios having wildly different start dates and potentially crashing the game"

#### Purpose of change

Fixes #45901

#### Describe the solution

Currently if you use "custom_initial_date" with a scenario all other properties are initialized randomly which is especially problematic with year having a max of 9999 which allows the game turn counter to overflow up to 73 times and crashes the game in case it is <0.

I don't think having all properties random is a good default. Instead custom_initial_date will be `"hour": 8, "day": 0, "season": "SPRING", "year": 0` by default so if a scenario had `"custom_initial_date": { "season": "winter" }` means first day of winter in year 0 at 8:00am

If you want to have a scenario start at a random point of time you can use -1, for example { "day": -1 } means a random day between 0 and season_length -1.

I also limited the roll for years to 0-10. Even with slower evolution times I don't think there will not be much difference after year 10.

And finally: custom scenarios will respect the spawn delay option.

#### Describe alternatives you've considered

Sadly at this point there is no randomization for season. The reason being that seasons are represented by an enum and I really do not want to add a "random" season.

#### Testing

Started new characters with scenarios "Ambush", "Next Summer" and "Evacuee" to see if start day matches my expectation (and the value shown when scrolling through the scenario list during character creation)

### CODE REVIEW

**Comment by Tamiore** [comment]
I do think all current scenarios that start "in winter" should always start:
1) during the winter of year 1. Random year does not seem intended in any of those scenarios. 
2) either on day 1 or winter or at least before 50% of the season has passed (otherwise we can get things like "starting 1 day from spring" which effectively makes it a spring start instead)

**Comment by satheon49** [comment]
I think it makes sense. I will change it so that the default values for custom_initial_date are `"time": 8, "day": 0, "season": "SPRING", "year": 0` unless overridden by the scenario.

As for random year it won't be used by any existing scenarios. Though I did want to leave the door open if someone wanted to make a walking dead style coma scenario.

**Comment by ZhilkinSerg** [comment]
We can change `year` field to use pair of numbers, so random value could be selected in-between (no necessary in current PR).

**Comment by satheon49** [comment]
If someone needs a random range for hour/day/seaons/year we can add it later. I think for now we should just wait if someone actually needs the random feature at all.

I updated the PR description and changed the behavior as described having `"hour": 8, "day": 0, "season": "SPRING", "year": 0` as fallbacks for any custom value not set by the scenario. Absence of `custom_initial_date` still means world gen settings.

Attention: With these settings it is possible to start a scenario **BEFORE** the cataclysm. Such a scenario still works, but zombies won't start evolving until the cataclysm happens (as defined by INITIAL_DAY in world gen)

File: src/game.cpp [hunk 1]
```diff
@@ -12051,13 +12051,17 @@ void game::autosave()
 
 void game::start_calendar()
 {
-    // Configured starting date overridden by scenario, calendar::start is left as Spring 1
```
Comment (Zireael07) [inline src/game.cpp]: why remove the comment?

Comment (satheon49) [reply src/game.cpp]: It got accidentally removed when I collapsed the two branches. I readded it.

File: src/scenario.cpp [hunk 2]
```diff
@@ -438,45 +441,46 @@ bool scenario::custom_initial_date() const
...
 
-bool scenario::random_initial_day() const
+bool scenario::random_day() const
 {
-    return _random_initial_day;
+    return _initial_day == -1;
 }
 
-bool scenario::random_initial_season( )const
+bool scenario::random_year() const
```
Comment (AMurkin) [inline src/scenario.cpp:454]: Maybe call them `is_random_...`? This will make it clearer, imo.

Comment (satheon49) [reply src/scenario.cpp]: Sure.

**Pull request merged**